### PR TITLE
fix - Fix webpack 4 hmr after refresh + fix rebuild performance + fix webpack 5 serialization

### PR DESCRIPTION
**Pull request by bebraw**
The issue is explained fully at https://github.com/storybookjs/storybook/issues/15233 .

There's a subtle difference between the new and old code. In my personal testing I found that restoring the old code path for webpack 4 fixes the issue so I did that in this PR. I also dropped a redundant dependency as `webpack-sources` isn't used anywhere.

This also fixes rebuild performance for both webpack 4 and 5.

In addition, https://github.com/storybookjs/storybook/issues/15702 is fixed (added missing serialization logic as it looks like there's no good default there).

Can you publish a canary version for @yannbf to test?

### CODE REVIEW

**Comment by bebraw** [comment]
I included #50 to this one so it's easier to merge.

File: src/plugin.ts [hunk 1]
```diff
@@ -191,6 +198,30 @@ export default class DocgenPlugin implements webpack.WebpackPluginInstance {
 
             const nameForCondition = module.nameForCondition() || "";
 
+            // Ignore already built modules for webpack 5
+            if (!compilation.builtModules.has(module)) {
```
Comment (7rulnik) [inline src/plugin.ts:202]: @bebraw Hmm, in PR description you said that `It's important not to process already built modules`, but there we return early for un-build modules. Am I missing something?

**Comment by bebraw** [comment]
I got the description wrong. It should be inverse as otherwise it won't work (tests will fail).

File: src/plugin.ts (hunk 1 above)
Comment (bebraw) [reply src/plugin.ts:202]: Here's the related API refactor: https://github.com/webpack/webpack/pull/7933 .

The interesting thing is that the API isn't used much. Here's the usage in webpack itself: https://github.com/webpack/webpack/search?q=builtModules&type= .

Webpack API is confusing here as it's not well documented what exactly happens on a recompilation.

**Comment by shilman** [comment]
@hipstersmoothie do you have an issue with us forking the repo? the merges/releases are happening too slowly for storybook's needs

**Pull request merged**